From: Philipp Zabel <p.zabel@pengutronix.de>
Date: Thu, 11 Oct 2018 16:36:07 +0200
Subject: [PATCH] media: coda: improve bitstream payload ready check

The bitstream prefetch unit reads data in 256 byte blocks with some kind
of queueing. For the decoder to see data up to a desired position in the
next run, the bitstream has to be filled for 2 256 byte blocks past the
desired position aligned up to the next 256 byte boundary.

Signed-off-by: Philipp Zabel <p.zabel@pengutronix.de>
---
 drivers/media/platform/coda/coda-common.c | 18 ++++--------------
 drivers/media/platform/coda/coda.h        | 12 ++++++++++++
 2 files changed, 16 insertions(+), 14 deletions(-)

diff --git a/drivers/media/platform/coda/coda-common.c b/drivers/media/platform/coda/coda-common.c
index 82fde20b71a1..160ff37566bd 100644
--- a/drivers/media/platform/coda/coda-common.c
+++ b/drivers/media/platform/coda/coda-common.c
@@ -1392,7 +1392,6 @@ static int coda_job_ready(void *m2m_priv)
 	if (ctx->inst_type == CODA_INST_DECODER && ctx->use_bit) {
 		int num_metas = ctx->num_metas;
 		struct coda_buffer_meta *meta;
-		unsigned int meta_size;
 		unsigned int payload = coda_get_bitstream_payload(ctx);
 		unsigned int count;
 
@@ -1422,21 +1421,12 @@ static int coda_job_ready(void *m2m_priv)
 
 		meta = list_first_entry(&ctx->buffer_meta_list,
 					struct coda_buffer_meta, list);
-		meta_size = meta->end - meta->start;
-		if (!stream_end && payload < meta_size + 512) {
+		if (!coda_bitstream_can_fetch_past(ctx, meta->end) &&
+		    !stream_end) {
 			trace_coda_not_ready(ctx, stream_end, src_bufs, num_metas, payload);
 			coda_dbg(1, ctx,
-				 "not ready: need 3*256 bytes beyond first buffer in bitstream (%d %d)\n",
-				 meta_size, payload);
-			return 0;
-		}
-
-		if (!src_bufs && !stream_end &&
-		    (payload < 512)) {
-			trace_coda_not_ready(ctx, stream_end, src_bufs, num_metas, payload);
-			coda_dbg(1, ctx,
-				 "not ready: not enough bitstream data (%d).\n",
-				 payload);
+				 "not ready: not enough bitstream data to read past %u (%u)\n",
+				 meta->end, ctx->bitstream_fifo.kfifo.in);
 			return 0;
 		}
 	}
diff --git a/drivers/media/platform/coda/coda.h b/drivers/media/platform/coda/coda.h
index f5918e8e5f0e..12175a986d20 100644
--- a/drivers/media/platform/coda/coda.h
+++ b/drivers/media/platform/coda/coda.h
@@ -311,6 +311,18 @@ static inline unsigned int coda_get_bitstream_payload(struct coda_ctx *ctx)
 	return kfifo_len(&ctx->bitstream_fifo);
 }
 
+/*
+ * The bitstream prefetcher needs to read at least 2 256 byte periods past
+ * the desired bitstream position for all data to reach the decoder.
+ */
+static inline bool coda_bitstream_can_fetch_past(struct coda_ctx *ctx,
+						 unsigned int pos)
+{
+	return (int)(ctx->bitstream_fifo.kfifo.in - ALIGN(pos, 256)) > 512;
+}
+
+bool coda_bitstream_can_fetch_past(struct coda_ctx *ctx, unsigned int pos);
+
 void coda_bit_stream_end_flag(struct coda_ctx *ctx);
 void coda_bit_debug_timeout(struct coda_ctx *ctx);
 
